Block/Service/AbstractCategoriesBlockService.php

```diff
@@ -0,0 +1,160 @@
+
+        $adminField = $this->getFormAdminType($formMapper, $this->getCategoryAdmin(), 'categoryId', 'category', array(
+            'label'              => 'form.label_category',
+        ), array(
+            'translation_domain' => 'SonataClassificationBundle',
+            'link_parameters'    => array(
+                array(
+                    'context' => $context,
+                ),
+            ),
+        ));
+
+        $formMapper->add('settings', 'sonata_type_immutable_array', array(
+            'keys' => array(
+                array('title', 'text', array(
+                    'label'    => 'form.label_title',
+                    'required' => false,
+                )),
+                array('context', 'choice', array(
+                    'label'    => 'form.label_context',
+                    'required' => false,
+                    'choices'  => $this->getContextChoices(),
+                )),
+                array($adminField, null, array(
+                )),
+            ),
+            'translation_domain' => 'SonataClassificationBundle',
+        ));
+    }
+    /**
+     * {@inheritdoc}
+     */
+    public function configureSettings(OptionsResolver $resolver)
+    {
+        $resolver->setDefaults(array(
+            'title'      => 'Categories',
+            'category'   => false,
+            'categoryId' => null,
+            'context'    => 'default',
+            'template'   => 'SonataClassificationBundle:Block:base_block_categories.html.twig',
+        ));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function load(BlockInterface $block)
+    {
+        if (is_numeric($block->getSetting('categoryId', null))) {
+            $block->setSetting('categoryId', $this->getCategory($block->getSetting('categoryId', null)));
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function prePersist(BlockInterface $block)
+    {
+        $id = is_object($block->getSetting('categoryId')) ? $block->getSetting('categoryId')->getId() : null;
+
+        $block->setSetting('categoryId', $id);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function preUpdate(BlockInterface $block)
+    {
+        $id = is_object($block->getSetting('categoryId')) ? $block->getSetting('categoryId')->getId() : null;
+
+        $block->setSetting('categoryId', $id);
+    }
+
+    /**
+     * @param $id
+     * @param $default
+     *
+     * @return CategoryInterface
+     */
+    protected function getCategory($id, $default = false)
+    {
+        if (is_object($id)) {
+            return $id;
+        } else {
+            if (!is_null($id) && $id) {
+                return $this->getCategoryManager()->find($id);
+            }
+        }
+
+        return $default;
+    }
+
+    /**
+     * @return CategoryManagerInterface
+     */
+    protected function getCategoryManager()
+    {
+        return $this->container->get('sonata.classification.manager.category');
```

Comment:
It's a bad practice to inject the whole container, only needed services is recommanded